mipengine/node/monetdb_interface/monet_db_connection.py

```diff
@@ -0,0 +1,99 @@
+from contextlib import contextmanager
+from typing import List
+
+import pymonetdb
+
+from mipengine import config
+from mipengine.common.node_catalog import node_catalog
+
+OCC_MAX_ATTEMPTS = 50
+
+
+class Singleton(type):
+    """
+    Copied from https://stackoverflow.com/questions/6760685/creating-a-singleton-in-python
+    """
+
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+class MonetDB(metaclass=Singleton):
+    """
+    MonetDB is a Singleton class because we want it to be initialized at runtime.
+
+    If the connection is a public module variable, it will be initialized at import time
+    from Celery and all the Celery workers will use the same connection instance.
+
+    We want one MonetDB connection instance per Celery worker/process.
+
+    """
+
+    def __init__(self):
+        global_node = node_catalog.get_global_node()
+        if global_node.nodeId == config.node.identifier:
+            node = global_node
+        else:
+            node = node_catalog.get_local_node(config.node.identifier)
+        monetdb_hostname = node.monetdbHostname
+        monetdb_port = node.monetdbPort
+        self._connection = pymonetdb.connect(
+            username=config.monetdb.username,
+            port=monetdb_port,
+            password=config.monetdb.password,
+            hostname=monetdb_hostname,
+            database=config.monetdb.database,
+        )
+
+    @contextmanager
+    def cursor(self):
+        try:
+            cur = self._connection.cursor()
+            yield cur
+        except Exception as exc:
+            raise exc
+        finally:
+            cur.close()
+
+    def execute_with_result(self, query: str) -> List:
+        """
+        Used to execute select queries that return a result.
+
+        Should NOT be used to execute "CREATE, DROP, ALTER, UPDATE, ..." statements.
+        """
+
+        # DO NOT Remove. The connection needs to be committed so it is up to date.
+        self._connection.commit()
```

Comment:
it makes no sense that you have to commit before executing. Looks like it is needed in cases where some previous action was not committed, but this way of dealing with this case seems not ideal. At least put a comment describing why "the connection needs to be committed so it is up to date..."

Reply:
You are right i will put a more descriptive comment.